subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/IncrementalFiler.java

```diff
@@ -16,41 +16,89 @@
 
 package org.gradle.api.internal.tasks.compile.processing;
 
+import com.google.common.collect.Sets;
+import org.gradle.api.internal.tasks.compile.incremental.processing.AnnotationProcessingResult;
+
 import javax.annotation.processing.Filer;
 import javax.annotation.processing.Messager;
 import javax.lang.model.element.Element;
+import javax.lang.model.element.PackageElement;
+import javax.lang.model.element.TypeElement;
 import javax.tools.Diagnostic;
 import javax.tools.FileObject;
 import javax.tools.JavaFileManager;
 import javax.tools.JavaFileObject;
 import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
 
 /**
  * A decorator for the {@link Filer} which ensures that incremental
  * annotation processors don't break the incremental processing contract.
  */
 abstract class IncrementalFiler implements Filer {
     private final Filer delegate;
+    private final AnnotationProcessingResult result;
     private final Messager messager;
 
-    IncrementalFiler(Filer delegate, Messager messager) {
+    IncrementalFiler(Filer delegate, AnnotationProcessingResult result, Messager messager) {
         this.delegate = delegate;
+        this.result = result;
         this.messager = messager;
     }
 
     @Override
     public final JavaFileObject createSourceFile(CharSequence name, Element... originatingElements) throws IOException {
-        checkOriginatingElements(name, originatingElements, messager);
+        recordGeneratedType(name, originatingElements, messager);
         return delegate.createSourceFile(name, originatingElements);
     }
 
     @Override
     public final JavaFileObject createClassFile(CharSequence name, Element... originatingElements) throws IOException {
-        checkOriginatingElements(name, originatingElements, messager);
+        recordGeneratedType(name, originatingElements, messager);
         return delegate.createClassFile(name, originatingElements);
     }
 
-    protected abstract void checkOriginatingElements(CharSequence name, Element[] originatingElements, Messager messager);
+    private void recordGeneratedType(CharSequence name, Element[] originatingElements, Messager messager) {
+        String generatedType = name.toString();
+        Set<String> originatingTypes = getTopLevelTypeNames(originatingElements);
+        checkGeneratedType(generatedType, originatingTypes, messager);
+        result.addGeneratedType(generatedType, originatingTypes);
+    }
+
+    protected abstract void checkGeneratedType(String generatedType, Set<String> orignatingTypes, Messager messager);
+
+    private Set<String> getTopLevelTypeNames(Element[] originatingElements) {
+        if (originatingElements == null || originatingElements.length == 0) {
+            return Collections.emptySet();
+        }
+        if (originatingElements.length == 1) {
```

Comment:
Is this necessary? It can be merged to the for loop below.

Reply:
It's the usual optimization for "zero, one, many". Instead of creating an expensive set implementation, we create a less memory intensive singleton set when we only have one element.